Add configure of quantization for dynamic graph

wanghaoshuang opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
New features


### PR changes
APIs

### Describe
Add configure of quantization for dynamic graph

原PR: https://github.com/PaddlePaddle/Paddle/pull/47530

###  和之前有什么不同

整个量化功能可以分为两部分：
- 基础设施：比较固定，主要体现为『往待量化模型中插入OP』的操作
- 算法实现：相对多变，主要体现为『转换/统计Tensor数值』的操作

本次调整，将量化功能和API放在合适的位置：

之前，
- 实现：『基础设施』和『算法实现』均实现在Paddle核心框架中，
-  API：Paddle核心框架不对外开放任何量化API，PaddleSlim对外开放所有量化API

现在，
- 整理优化现有量化功能实现
- 实现：基础设施在Paddle核心框架中，算法实现在PaddleSlim中
- API：Paddle核心框架对外开放『基础设施』API, PaddleSlim对外提供『算法实现』API



### 当前PR为本次调整的部分工作

当前PR包括以下架构图中的QuantConfig, QuanterFactory和部分quanter
![image](https://user-images.githubusercontent.com/7534971/201936175-b1128176-b22a-483c-8f93-69ed88f4336a.png)


#### 关键概念

##### Quanter

对OP的输入或输出执行量化操作。
  -  FakeQuanter：执行量化和反量化
  -  QuanterWithObserver：除了Quanter本身的职责，还附带Observer的职责

![image](https://user-images.githubusercontent.com/7534971/201936685-c41e0eb9-3d91-4f4d-b419-cc120597b9f1.png)

##### ObserverFactory & QuanterFactory

  - 信息载体：是量化配置信息的主要载体，存储类型和参数
  - 构造实例：用于构造Observer/Quanter实例
  
##### QauntConfig
对ObserverFactory & QuanterFactory实例进行管理，存储模型量化配置的所有信息。

#### 关键类设计

在执行量化操作之前，用户首先要配置量化相关的信息，主要是指定每层的各个输入使用什么Observer或Quanter.
配置功能主要涉及三个类：QuantConfig，SingleLayerConfig，ObserverFactory/QuanterFactory。以上三个类分别存储整个模型、单层和单个Observer的配置信息。他们之间是组合关系，QuantConfig实例中包含若干SingleLayerConfig实例，SingleLayerConfig实例包含若干ObserverFactory/QuanterFactory实例。

![image](https://user-images.githubusercontent.com/7534971/201937129-47cb646a-a88a-422f-8fb1-bea62bd8e35c.png)

wanghaoshuang committed:
Add configure of quantization for dynamic graph

wanghaoshuang committed:
Add comments

wanghaoshuang committed:
Add comments

wanghaoshuang committed:
Install quantization package

wanghaoshuang committed:
Fix six

wanghaoshuang committed:
Remove fluid

wanghaoshuang committed:
Fix sample codes

wanghaoshuang committed:
Fix typo

wanghaoshuang commented:
## 从Paddle API角度看，和之前有什么不同

整个量化功能可以分为两部分：
- 基础设施：比较固定，主要体现为『往待量化模型中插入OP』的操作
- 算法实现：相对多变，主要体现为『转换/统计Tensor数值』的操作

本次调整，将量化功能和API放在合适的位置：

之前，
- 『基础设施』和『算法实现』均实现在Paddle核心框架中，API不在Paddle核心框架对外开放
- PaddleSlim调用Paddle核心框架的量化模块，并在PaddleSlim对外开放量化API

现在，
- 整理优化现有量化功能实现
- 基础设施在Paddle核心框架中，算法实现在PaddleSlim中
- Paddle核心框架对外开放『基础设施』API, PaddleSlim对外提供『算法实现』API

yghstill commented on python/paddle/quantization/quanters/abs_max.py:
```diff
@@ -0,0 +1,151 @@
+    def quant_axis(self):
+        return 0
+
+    def scales(self):
```
这里的scales为什么return None，作用是获取self._scale的值吗？

wanghaoshuang replied:
Fixed. And add some comments for BaseQuanter.

wanghaoshuang committed:
Remove TRTConfig and fix absmax_quanter

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into quant_config

wanghaoshuang committed:
Fix code style by isort hook

chenwhql reviewed (dismissed):
LGTM for setup.py.in

qingqing01 commented on python/paddle/quantization/config.py:
```diff
@@ -0,0 +1,441 @@
+         lower than `add_layer_config`.
+
+         Args:
+             layer(Union[str, list]): One or a list of layers' full name.
```
In line 138, it is group.  layer -> group

qingqing01 commented on python/paddle/quantization/config.py:
```diff
@@ -0,0 +1,441 @@
+        and `add_prefix_config`.
+
+        Args:
+            layer(Union[str, list]): One or a list of layers' type.
```
layer -> layer_type.  also give some example, what is layers'  type.

qingqing01 commented on python/paddle/quantization/config.py:
```diff
@@ -0,0 +1,441 @@
+            from paddle.quantization.quanters import FakeQuanterWithAbsMaxObserver
+            quanter = FakeQuanterWithAbsMaxObserver(moving_rate=0.9)
+            q_config = QuantConfig(activation=None, weight=None)
+            q_config.add_qat_layer_mapping(Conv2D, Conv2D)
```
This example is not easy to understand the function.  Maybe change `target` layer to QuantizedConv2D or one customized Layer ?

qingqing01 commented on python/paddle/quantization/factory.py:
```diff
@@ -0,0 +1,133 @@
+       .. code-block:: python
+
+            # Given codes in ./custom_quanter.py
+            from paddle.quantization.factory import quanter
```
the namespace factory may not export to users

qingqing01 commented on python/paddle/quantization/quanters/abs_max.py:
```diff
@@ -0,0 +1,152 @@
+    :math:`scale = (moving\_rate*accum+max(abs(x)))/(moving\_rate*state+1)`
+    :math:`range = 2^{bit\_length - 1} - 1`
+    :math:`Out = round(X / scale * range) * scale / range`
+
```
Add docs later

wanghaoshuang replied:
Done

qingqing01 commented on python/paddle/tests/test_quant.py:
```diff
@@ -0,0 +1,151 @@
+# copyright (c) 2022 paddlepaddle authors. all rights reserved.
```
If will add more tests for quantization,  can create a new directory

jeff41404 commented on python/paddle/quantization/__init__.py:
```diff
@@ -12,35 +12,43 @@
 # limitations under the License.
 
-from ...fluid.contrib.slim.quantization.imperative.ptq_config import (
+from ..fluid.contrib.slim.quantization.imperative.ptq_config import (
```
If these classes/functions `ptq_config`(and below) need to be used, it is recommended to migrate their implementation here, and the fluid is also cleaning.

wanghaoshuang replied:
`Fluid` will be removed in another PR: https://github.com/PaddlePaddle/Paddle/pull/48717

jeff41404 replied:
ok, thanks

wanghaoshuang committed:
Refine the code and docs
1. Fix typo
2. Create new directory for unitests of quantization
3. Import quanter to paddle.quantization package
4. Refine docs

jeff41404 commented on python/paddle/quantization/__init__.py:
```diff
@@ -12,35 +12,47 @@
+from .factory import quanter
+from .quanters import FakeQuanterWithAbsMaxObserver
+
+__all__ = []
+__all__ += ["QuantConfig"]
+__all__ += ["BaseQuanter"]
+__all__ += ["quanter"]
+__all__ += ["FakeQuanterWithAbsMaxObserver"]
```
it's better to use one python list to represent `__all__` instead of merging many lists with `+=`, more intuitive and better monitoring.

jeff41404 replied:
`QuantConfig`, `BaseQuanter`, `quanter` and `FakeQuanterWithAbsMaxObserver` also in other file's `__all__`, this is prohibited, which will cause the same API to have two recommended paths for users with inconsistent usage and easy confusion.

wanghaoshuang replied:
Thx. Done.

jeff41404 commented on python/paddle/quantization/quanters/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+from .abs_max import FakeQuanterWithAbsMaxObserver
+
+__all__ = []
+__all__ += ["FakeQuanterWithAbsMaxObserver"]
```
same issue as python/paddle/quantization/__init__.py:L54

wanghaoshuang committed:
1. Implement FakeQuanterWithAbsMaxObserver by extending QuanterFactory to make the doc can be extracted from code
2. remvoe some __all__

wanghaoshuang committed:
Add unitest for customized quanter

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into quant_config

wanghaoshuang committed:
1. Fix smaple codes
2. remove unused __all__

jeff41404 reviewed (approved):
LGTM

Ligoml commented on python/paddle/quantization/config.py:
```diff
@@ -0,0 +1,444 @@
+             weight(QuanterFactory): Quanter used for weights.
+
+         Examples:
+        .. code-block:: python
```
.. code-block:: python前面需要再加一段缩进，参考格式：
```
    Examples:
        .. code-block:: python

            import paddle

            x = paddle.to_tensor([2, 3, 4], 'float64')
            y = paddle.to_tensor([1, 5, 2], 'float64')
            z = paddle.add(x, y)
            print(z)  # [3., 8., 6. ]
```

wanghaoshuang replied:
Thx. 后续另起PR修复。

Ligoml commented on python/paddle/quantization/factory.py:
```diff
@@ -0,0 +1,131 @@
+    Annotation to declare a factory class for quanter.
+
+    Args:
+        class_name (str) - The name of factory class to be declared.
```
```suggestion
        class_name (str): The name of factory class to be declared.
```

Ligoml reviewed (approved):
api文档有一些小问题，可以先合入，再补充中文文档

XiaoguangHu01 reviewed (approved):
LGTM
